peps/pep-0790.rst

```diff
@@ -0,0 +1,400 @@
+            group_tb = group.__traceback__
+            combined_tb = _combine_tracebacks(parent_tb, group_tb)
+            for exc in group.exceptions:
+                if isinstance(exc, BaseExceptionGroup):
+                    _flatten(exc, combined_tb)
+                else:
+                    tb = _combine_tracebacks(combined_tb, exc.__traceback__)
+                    result.append(copy.copy(exc).with_traceback(tb))
+
+        _flatten(eg)
+        return result
+
+
+    def _combine_tracebacks(
+        tb1: TracebackType | None,
+        tb2: TracebackType | None,
+    ) -> TracebackType | None:
+        """
+        Combine two tracebacks, putting tb1 frames before tb2 frames.
+
+        If either is None, return the other.
+        """
+        if tb1 is None:
+            return tb2
+        if tb2 is None:
+            return tb1
+
+        # Convert tb1 to a list of frames
+        frames = []
+        current = tb1
+        while current is not None:
+            frames.append((current.tb_frame, current.tb_lasti, current.tb_lineno))
+            current = current.tb_next
+
+        # Create a new traceback starting with tb2
+        new_tb = tb2
+
+        # Add frames from tb1 to the beginning (in reverse order)
+        for frame, lasti, lineno in reversed(frames):
+            new_tb = types.TracebackType(
+                tb_next=new_tb, tb_frame=frame, tb_lasti=lasti, tb_lineno=lineno
+            )
+
+        return new_tb
+
+
+.. raw:: html
+
+   </details>
+
+
+Adding ``BaseException.preserve_context()``
+-------------------------------------------
+
+A new method ``preserve_context()`` will be added to ``BaseException``, with the
+following signature:
+
+.. code-block:: python
+
+    def preserve_context(self) -> contextlib.AbstractContextManager[Self]:
+        """
+        Context manager that preserves the exception's __context__ and
+        __suppress_context__ attributes.
+
+        When entering the context, the current values of __context__ and
+        __suppress_context__ are saved. When exiting, these saved values are
+        restored to the exception. This allows raising an exception inside
+        an except block without changing its context chain.
+        """
+
+Usage example:
+
+.. code-block:: python
+
+    # We're an async web framework, where user code can raise an HTTPException
+    # to return a particular HTTP error code to the client.  However, it may
+    # (or may not) be raised inside a TaskGroup, so we need to use `except*`;
+    # and if there are *multiple* such exceptions we'll treat that as a bug.
+    try:
+        user_code_here()
+    except* HTTPException as group:
+        first, *rest = group.flat_exceptions()
+        if rest:
+            raise  # handled by internal-server-error middleware
+        ...  # logging, cache updates, etc.
+        with first.preserve_context():
+            raise first
```

Comment:
Maybe more a question for Discuss, but: If the usage is always going to be `with e.preserve_context(): raise e`, does it make more sense to make a single method that does both of these things? For example, a method `e.raise_with_preserved_context()`.